src/preact/base-element.js

```diff
@@ -188,6 +193,28 @@ export class PreactBaseElement extends AMP.BaseElement {
     this.scheduleRender_();
   }
 
+  /**
+   * @return {!API_TYPE}
+   * @protected
+   */
+  api() {
+    return devAssert(this.ref_.current);
+  }
+
+  /**
+   * @param {string} alias
+   * @param {function(!API_TYPE, !../service/action-impl.ActionInvocation)} handler
+   * @param {../action-constants.ActionTrust} minTrust
+   * @protected
+   */
+  registerApiAction(alias, handler, minTrust) {
```

Comment:
would we make this a no-op in the bento mode envelope?

Reply:
I don't think so. We still need to allow action execution. E.g. a caller should be able to call `carousel.goToSlide(0)` no matter what mode we are in.

Reply:
Does this mean we would want a Bento version of `amp-state`? I don't recall having discussions about this

Reply:
No need for Bento `amp-state`. But the on-actions are not part of `amp-state` - that's just normal imperative API for an element. Just like `video.pause()` - there has to be a way to pause a video programmatically. Otherwise the usability of an element is significantly reduced. This is also usually not a big deal to support - it'd most often be a simple passthrough `element.goToSlide -> api(component).goToSlide`.

Reply:
To clarify, `amp-state` might use element on-actions, but these actions exist independent of `amp-state`.

Reply:
I don't follow. In Bento mode users would still be using `<amp-base-carousel>` on the page, whereas in Preact mode users would be using `<BaseCarousel>`. My understanding is the former still requires `amp-state` interop and the latter can call apis directly on the Preact element instance in the virtual DOM. Am I getting this right?

Reply:
Yes. That's correct. Except 

> former still requires amp-state interop

I was just clarifying that on-action support is related, but not the same as `amp-state` interop. Thus, even if we don't expect users to use `amp-state` in Bento mode, they should still be able to call on-actions.

Reply:
Oh I see, so `on=` should be supported in `PreactBaseElement` rather than imported independently?

Reply:
Not certain what you mean by "rather than imported independently". But, yes, imperative APIs could be very important for a component's usability. It's rather similar to, say, an attribute mutation but for concepts that are not easy to express via DOM markup.

Reply:
And one more clarification. I also don't expect Bento mode to use `on="tap:carousel.goToSlide"` type of thing. But somehow the element needs to export an API to make it possible still or the element's API surface will be incomplete. Currently it's possible with `element.enqueAction('goToSlide')` - a really complicated way to do it, but sufficient for testing. We will be looking into how we can make these API calls more ergonomic in Bento world.

Reply:
> API calls more ergonomic in Bento world

This PR sets the stage for it! One potential option would be to forward all keys within the `.getApi()` object to the element

Reply:
Thanks for clarifying, this all makes sense and I think the `imported independently` was groggy Monday brain. Feel free to disregard.

Reply:
I think forwarding is exactly what we should do. But Not yet 100% certain it'd be that simple. Another option is to simply export `api` property on the element. E.g. users would call `element.api.goToSlide()`. That has two benefits:

1. We could give `.api` more clear async semantics. It's not available until the element is upgraded.
2. We don't ever have to worry about method collision with any DOM APIs - it's basically functions as a namespace.

But it has one big downside: per Preact semantics the imperative handle can easily change, which is not something a Bento user would always expect. `element.api@time1 !== element.api@time2`.

Thus, I think, some sort of proxying is a more likely solution for us. This is something to fully design and TODO, but not subject for this PR specifically since it's only taking care of the element->preact bridge.

Reply:
I created https://github.com/ampproject/wg-bento/issues/36 to track the element API question.